### PR TITLE
Fixes for GatherND, Multinomial

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR registers the cuda/rocm kernel for GatherND for bool type and exports the Multinomial op as Aten op instead of the onnx op since it is missing a CUDA kernel in ORT.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -966,6 +966,43 @@ def run_step(model, input):
...
+
+        def forward(self, input):
+            return torch.multinomial(input, self.num_samples, self.replacement)
+
+    torch.manual_seed(5032)
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    device = 'cuda'
+    pt_model = NeuralNetDiagonal(num_samples, replacement).to(device)
+    dbo = DebugOptions(save_onnx = True, onnx_prefix = 'ort_aten_Mult')
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:992]: clean up the dbo before merge?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -966,6 +966,43 @@ def run_step(model, input):
...
+            super(NeuralNetDiagonal, self).__init__()
+            self.num_samples = num_samples
+            self.replacement = replacement
+
+        def forward(self, input):
+            return torch.multinomial(input, self.num_samples, self.replacement)
+
+    torch.manual_seed(5032)
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:989]: is torch.backends.cudnn.benchmark related to multinomial kernel?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -966,6 +966,43 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
         _test_helpers.assert_values_are_close(ort_input.grad, pt_input.grad)
 
+# Since multinomial is a generator function, we do not have to test for gradient
+# Two consecutive calls on the torch.multinomail on a probability distribution with more
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:970]: calling torch.manual_seed will reset the torch generator, the resulting generated sequence will be the same. 


>>> torch.manual_seed(42)
<torch._C.Generator object at 0x7f1d20c0f990>
>>> torch.multinomial(x,2,True)
tensor([0, 0])
>>> torch.multinomial(x,2,True)
tensor([1, 0])
>>> torch.multinomial(x,2,True)
tensor([2, 2])
>>> torch.multinomial(x,2,True)
tensor([3, 3])
>>> torch.manual_seed(42)
<torch._C.Generator object at 0x7f1d20c0f990>
>>> torch.multinomial(x,2,True)
tensor([0, 0])
>>> torch.multinomial(x,2,True)
tensor([1, 0])
>>> torch.multinomial(x,2,True)
tensor([2, 2])
>>> torch.multinomial(x,2,True)
tensor([3, 3])

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 4]
```diff
@@ -78,6 +78,12 @@ def diagonal(g, self, offset, dim1, dim2):
     return g.op("com.microsoft::ATenOp", self, offset, dim1, dim2,
                 name_s='aten::diagonal')
 
+@register_symbolic('multinomial')
+def multinomial(g, self, num_samples, replacement=False, generator=None):
+    if generator is not None and not sym_help._is_none(generator):
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:83]: should the condition be 'and' or 'or'

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py [hunk 5]
```diff
@@ -125,3 +132,5 @@ def adaptive_avg_pool2d_gradient():
         (('ATenOp', 'com.microsoft'), ['GO(0)', 'I(0)'], [
          'GI(0)'], {'name': {'value': 'aten::_adaptive_avg_pool2d_backward', 'dtype': 'string'}}),
     ]
+
+CustomGradientRegistry.register_custom_stop_gradient_edges([0], 'com.microsoft', 'ATenOp', 'aten::multinomial', '')
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:136]: Vincent, please share your thoughts on using this approach to specific the non-differentiable edges. 

Another design option is to introduce an attribute '_differentiable_inputs', with value being the indexes of differentiable inputs.

Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:136]: I actually think that for this specific case here, using registry_stop_edges is better than using attribute. Because this is a schema-level property. Using attribute is more meaningful if it's possible that different nodes of same Op type can have different attribute values. But for the case here the 0th input is always non-differentiable. Using attribute also requires extra work to set the attribute value when generating forward graph. It might be simple for current custom symbolic registration, but exporter may not happy to take this work in the future.

But do we really need to add such register_stop_edges for now? Actually we can stop to build gradient for this ATenOp when looking into the output type of this Op. It's only 1 output with INT64 dtype. We can always ignore such nodes. We can ignore nodes if none of the outputs is decimal type.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 3 above)
Comment (ashbhandare) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:970]: that is right, are you suggesting we reset the seed between calls to the ort and torch models, and still pass random data?

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py (hunk 4 above)
Comment (ashbhandare) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:83]: this is following https://github.com/pytorch/pytorch/blob/bbe25af0df2f2f6a58f6cc4ee5899efcba2ee336/torch/onnx/symbolic_opset9.py#L2879

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 2 above)
Comment (ashbhandare) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:989]: It shouldn't be required for this test. removing

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 3 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:970]: yes

File: orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py (hunk 5 above)
Comment (ashbhandare) [reply orttraining/orttraining/python/training/ortmodule/_custom_gradient_registry.py:136]: I think adding the condition of not following non-decimal edges can be used to remove may ops like Shape, Size, GatherND etc from the STOP_GRADIENT_EDGES list. But since that condition alone is not enough, adding the same mechanism for Aten ops allows us to keep the gradient building logic consistent. Additionally, this allows us to have a mechanism for future aten ops where just the INT64 edge condition is not enough.
We can consider refactoring the gradient builder to add this rule of INT64 in a separate PR if necessary.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 6]
```diff
@@ -966,6 +966,44 @@ def run_step(model, input):
...
+        # reset manual seed to reset the generator
+        torch.manual_seed(5032)
+        prediction = model(input)
+        return prediction
+
+    pt_input = torch.rand(input_shape, dtype=torch.float, device=device)
+    ort_input = copy.deepcopy(pt_input)
+    pt_prediction = run_step(pt_model, pt_input)
+    ort_prediction = run_step(ort_model, ort_input)
+    # run the ort prediction again since the first call involves export
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1000]: ah... interesting...

File: orttraining/orttraining/core/graph/gradient_definition_registry.h [hunk 7]
```diff
@@ -51,8 +51,22 @@ class GradientDefinitionRegistry {
     definitions_.emplace(key, definition);
   }
 
+  void SetStopGradientEdgesForNode(const std::string& key, const std::unordered_set<size_t> edges) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_definition_registry.h:54]: pass by const ref

File: orttraining/orttraining/core/framework/gradient_graph_builder.h [hunk 8]
```diff
@@ -138,6 +138,8 @@ class GradientGraphBuilder {
   // Tracks tensors that are stashed in the forward pass for later use in backward pass.
   std::unordered_set<std::string> stashed_tensors_;
 
+  const std::unordered_set<size_t>* GetStopGradientEdges(const Node& node) const;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/gradient_graph_builder.h:141]: nit, comment here 
return index for stop gradient input edges if found. 
return nullptr if no stop gradient edge is not found
